### ISSUE TITLE
Performance issue when accessing an extremely large (10GB) longtensor

### ISSUE THREAD

**Issue by EvenOldridge** [pytorch/pytorch#20562]
## 🐛 Bug

I'm encountering a very strange performance issue when accessing an extremely large tensor.  I first encountered it when loading a whole dataset into gpu memory on a Tesla v100, which may be required for reproducibility because smaller tensors don't demonstrate the effect.

If the tensor is above a certain size, then once you index past a specific value, access to the tensor (or at least the performance of the dnn) begins to take longer and longer.  The linked repo includes an example notebook that demonstrates the performance dropoff.

There doesn't seem to be an issue with tensors in cpu memory.

Here are the things I've tested so far to help isolate:

 - Different tensor shapes (2x wider tensors results in degredation at ~6M samples)
 - Changed the tensor size to be within the limit (no slowdown occurs)
 - CPU Memory (No issues)
 - Different GPUs, although all on the same DGX-1 (hopefully validating hardware is functioning)
 - Artifically starting the dataloader at the point where it slows down. (Slowdown is immediate and performance is much worse (25K samples/s)
 - Breaking up the tensor into multiple blocks to see if being contiguous matters (multiple tensors don't have the same effect. Only one larger tensor demonstrates this issue)
 - Started with multiple blocks and concated together on the GPU. (issue shows up again)
 - Tested just the indexing of the single large random tensor to see if it was impacted (no slowdown)

This is possibly due to a int (or long) variable in the memory addressing of the tensor; If I calculate the point in the tensor where the slowdown occurs (45x4bytesx~12M = 2.16B) that's suspiciously close to the int limit of 2147483647.

What's strange is that the slowdown only occurs if there is significant access beyond that range. I tested a tensor that was 4 bytes larger (and one that was 100K larger) and neither of those displayed significant problems. It's only when it's much larger that it seems to cause the issue.

As mentioned above, for a larger tensor if I start accessing in that region the slowdown is immediate and much more pronounced.

It's worth noting I encountered a similar and more nefarious issue when shuffling by index a tensor loaded from dl_pack of this size where the region of the data beyond the int limit was all 0s.

I understand tensors of this size are unusual (at least for now) but it would be great to figure out the limitations so that others don't run into this and so that we can plan for future GPUs with much more RAM and in GPU memory datasets.

## To Reproduce

Run the notebook available in the repo: https://github.com/EvenOldridge/HugeTensor
The notebook uses ignite version 0.1.2 and contains a batch dataloader that I'm prototyping and hope to integrate into pytorch once its kinks are worked out.

## Expected behavior

Performance of the nn should be consistent when indexing the same tensor.  As soon as the data is split into 3 tensors the performance is fine as expected.  Performance on the CPU for a single large tensor is also fine.  No performance dropoff should occur when accessing a single large tensor in GPU memory. 

## Environment

 - PyTorch Version (e.g., 1.0): 1.01
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: TeslaV100 32 Gig (on a DGX-1)
 - Any other relevant information:

**Comment by colesbury** [comment]
I'm having trouble running your example. The transitive dependencies seem to require a different CUDA version from what I have installed (CUDA 10). It would help if you can simplify the repro code and remove unnecessary dependencies.

> If the tensor is above a certain size, then once you index past a specific value, access to the tensor (or at least the performance of the dnn) begins to take longer and longer

We have some operations that are much slower on very large tensors (more than 2147483647 elements) because they usekernels with int64 index calculations instead of int32 (and may involve expensive int64 divisions).

However, that doesn't sound like what you are describing. It sounds like you are using a "normal" sized input that comes from slicing a very large Tensor. That won't trigger the above behavior.

> It's worth noting I encountered a similar and more nefarious issue when shuffling by index a tensor loaded from dl_pack of this size where the region of the data beyond the int limit was all 0s.

Can you file an issue specifically for this? (I'm not sure which shuffle op you're referring to). We support operations on large tensors, but bugs tend to be more common because the code paths are used less often.

The "Testing tensor indexing" test in the notebook is operating on a CPU tensor:

```python
features = torch.LongTensor(15000000, 45).random_(0, 2**22)
features.cuda()
```

`features.cuda()` returns a copy on the GPU. It doesn't modify the Tensor in-place. You would need to write:

```python
features = torch.LongTensor(15000000, 45).random_(0, 2**22)
features = features.cuda()
```

or better (which avoids the CPU tensor entirely):

```python
features = torch.empty(15000000, 45, dtype=torch.int64, device='cuda').random_(0, 2**22)
```

**Comment by EvenOldridge** [comment]
Thanks for the follow up Sam.  I updated the repo to eliminate any unnecessary dependencies leaving only torch, ignite, and base python libraries.  I also rewrote the indexing test (thanks for the code) and ran it with similar effect.

Finally, I made another discovery while reviewing the code, which is that the issue only occurs if I shuffle the data.  I'm doing so using the following function:

    def shuffle(self):
        print('shuffling batch')
        idx = torch.randperm(self.num_samples, dtype=torch.int64, device='cuda')
        self.features = self.features[idx]
        self.targets = self.targets[idx]   

Which is what the examples I saw on this method showed.  Indexing is normally done by longtensor, so I don't see why that shuffle would cause an issue for data after the int limit, but if I remove the shuffle then the issue doesn't occur.  

As you can see from the notebook the throughput drops from 113K examples/s to 42K examples/s and continues to degrade if you run for more epochs.

Let me know if you have any issues reproducing.  I'll file the other issue as well, although it's more complex, involving dlpack.  It also had the same connection to shuffling.

**Comment by EvenOldridge** [comment]
One further follow up that I tested was to test a different type of shuffle (offset by 1) with the following function:

    def shuffle(self):
        print('shuffling batch')
        #idx = torch.randperm(self.num_samples, dtype=torch.int64, device='cuda')
        # offset by 1
        idx = torch.arange(1, self.num_samples+1, dtype=torch.int64, device='cuda')
        idx[self.num_samples-1] = 0
        self.features = self.features[idx]
        self.targets = self.targets[idx]     

I saw the same impact in the first epoch, with performance beginning to fall off once the integer limit of memory had been accessed.  The second and third epochs were much worse than the random shuffle at the start (25K examples/s) but performance increased over the epoch, peaking at 60K examples/s.

It seems like the mass update of such a large tensor by reindexing causes a significant slowdown in performance.
